src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazy`1.cs

```diff
@@ -574,7 +574,11 @@ public void CompleteFromTask(Task<T> task)
                 }
                 else if (task.IsFaulted)
                 {
-                    this.TrySetException(task.Exception!);
+                    RoslynDebug.AssertNotNull(task.Exception);
+                    if (task.Exception.InnerExceptions.Count > 0)
+                        this.TrySetException(task.Exception.InnerExceptions[0]);
+                    else
+                        this.TrySetException(task.Exception);
```

Comment:
how did this manifest?  

note, the rest of the PR lgtm. But i def want @jasonmalinowski eyes on this bit to ensure he thinks it is ok.

Reply:
Seems odd that we would unwrap the exception here.

---
In reply to: [426124196](https://github.com/dotnet/roslyn/pull/44306#discussion_r426124196) [](ancestors = 426124196)

Reply:
`TrySetException` wraps the exception you provide it. If you provide it a wrapped exception, it will end up wrapped twice and subsequent `await` operations will misbehave.

Reply:
Should we only unwrap if it's just wrapping a single exception?  It seems a little weird that we could be wrapping multiple exceptions and then drop all but one of them

Reply:
The behavior here is the same behavior that async/await uses.

Reply:
@CyrusNajmabadi see here:
https://github.com/dotnet/runtime/blob/7d6a355059bf46994c921bdc1dbcfb162cddc713/src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/TaskAwaiter.cs#L176-L182

Reply:
Great. Could you add a comment/link to that effect in the code? 

That would satisfy me without jason's explicit signoff

Reply:
Added comment

Reply:
I admit I don't like the fact that we're dropping the other exceptions on the floor if there were multiple root causes. I'm not sure "that's what async/await" does to me is a strong argument since that behavior seems pretty suspect at first glance. ;-)

Reply:
Switching the code to `if (task.Exception.InnerExceptions.Count == 1)` seems 100% reasonable to me.

Reply:
Fixed. We now defer the discarding of exceptions to the point where this is awaited.